### PR TITLE
[vLLM] Enable torch.compile for sample_from_logits, gather_logprobs in SPMD mode + coherence tests

### LINKED ISSUE #3589: [vLLM] torch.compile on sample_from_logits produces incorrect output under SPMD/TP
## Problem

Compiling `sample_from_logits` with `torch.compile(backend="tt")` under SPMD tensor parallel mode causes all tokens to be token_id 0. The function works correctly when run eagerly under TP, and works correctly when compiled without TP.

The original code had a conditional that skipped compilation under TP with this TODO:
```python
# TODO: Under SPMD mode, sample_from_logits has correctness issue.
#       Re-enable the torch.compile once the issue is fixed in torchxla.
```

An attempt was made to remove this guard (reasoning that logits are already replicated before reaching sampling), but the issue persists even with explicit `sharding_constraint_tensor` replication before the compiled function.

## Reproduction

With `enable_tensor_parallel: True` on n300 (Llama-3.2-3B or Qwen3-0.6B):

1. Add `@torch.compile(backend="tt", fullgraph=True, dynamic=False)` to `sample_from_logits`
2. Run `test_greedy_determinism[n300]`
3. Output is `!!!!!!!!!!!!!!!!!!!!` (all token_id 0)
4. Remove the decorator → output is correct

## Current workaround

`sample_from_logits` is compiled conditionally — only when `enable_tensor_parallel` is False. This means TP mode has an eager fallback in the sampling hot path.

## What was ruled out

- **Logit sharding**: Adding unconditional `sharding_constraint_tensor(logits, mesh, (None, None))` before `sample_from_logits` did not fix the issue
- **Model-specific**: Reproduces with both Llama-3.2-3B and Qwen3-0.6B (both tied-embedding models)
- **`ParallelLMHead` vs `VocabParallelEmbedding`**: Both models use tied embeddings, so the `ParallelLMHead` sharding constraint path is never entered

## Expected fix

Investigate the interaction between `torch.compile(backend="tt")` and XLA SPMD partitioning in the sampling graph (`argmax`, `multinomial`, temperature scaling, top-k/top-p filtering). The compiled graph likely has incorrect sharding propagation for one of these ops.

### LINKED ISSUE #3663: [vLLM] Add output coherence smoke test to SamplingParams tests to catch gibberish regressions
### Problem
During the #3589 investigation (compiling `sample_from_logits` under SPMD), we discovered that the existing test suite can miss cases where the model produces gibberish output — deterministic and non-degenerate (passes all current assertions), but not coherent English. The degenerate token guard added to `test_greedy_determinism` catches all-identical tokens but not this subtler failure mode.

### Suggested approach
Add a lightweight coherence smoke test that gives the model a few prompts with obvious factual completions (e.g., "The capital of France is" → expects "Paris") and asserts the expected keyword appears in the greedy output. This provides a stronger regression guard than structural checks alone without being brittle.

A prompt evaluation script (`eval_coherence_prompts.py`) was prototyped on the `kmabee/issue_3589_spmd_debug` branch to find prompts that all three test models (OPT-125M, TinyLlama, Qwen3-0.6B) can reliably complete.

### Priority
Nice-to-have — the existing degenerate guard and visual inspection of `[TESTOUT]` lines catch the worst cases. This would close the gap for subtler corruption.

### LINKED ISSUE #3672: [TT backend] Experimental compile fails on second invocation under SPMD for sample_from_logits
### Problem

The experimental compile path (`_call_experimental_compile` using `torch.export` + `extract_compiled_graph`, enabled by default in #2996) fails when a compiled function is invoked a second time with different data under SPMD. In standalone repros this manifests as `Error code: 13`. In the vLLM plugin context it silently returns the warmup result.

This was discovered during #3589 when enabling `torch.compile` on `sample_from_logits` under SPMD/TP. The legacy compile path (`tt_legacy_compile=True`) does not have this issue.

### Reproduction

Under SPMD (`xr.use_spmd()`), compiling a pure function (no model parameters) with the experimental path:
```python
compiled_fn = torch.compile(fn, backend="tt", fullgraph=True, dynamic=False)
compiled_fn(torch.zeros(1, 32000).to(device))  # warmup — works
compiled_fn(torch.randn(1, 32000).to(device))  # Error code: 13
```

The HLO graph from the first call has 1 input and 1 output (confirmed via `XLA_DYNAMO_DEBUG=1`) — the input is correctly treated as a graph input, not constant-folded. But re-executing the cached graph with new data fails.

Does not reproduce:
- Without SPMD
- With legacy compile (`options={"tt_legacy_compile": True}`)
- When the function accesses model sub-modules (e.g. `gather_logprobs` works fine)

### Current workaround

`sample_from_logits` uses legacy compile via decorator option:
```python
@torch.compile(backend="tt", fullgraph=True, dynamic=False,
               options={"tt_legacy_compile": True})
def sample_from_logits(self, logits, sampling_metadata):
```

### Notes

- Debug scripts on branch `kmabee/issue_3589_spmd_debug`: `debug_constant_fold*.py`
- `gather_logprobs` uses the same warmup pattern but works — possibly because it references `self.sampler` sub-modules which changes the graph structure
- The `optimized_mod` in `dynamo_bridge.py` has SPMD-specific input sharding check logic (lines 589-604) that may be involved

### PR DESCRIPTION
**Pull request by kmabeeTT**
### Ticket
Closes #3589
Closes #3663
Closes #3672

### Problem description
- `sample_from_logits` was not compiled under SPMD/TP mode due to a correctness issue where all tokens became token_id 0. The function ran eagerly in TP mode, leaving a performance gap in the sampling hot path.

### What's changed
- **`model_runner.py`**: Enable `@torch.compile` on `sample_from_logits` and `gather_logprobs`. Create fresh dummy tensors per precompile iteration to avoid stale XLA tensor IDs under SPMD (#3672). Move logit replication from an external `sharding_constraint_tensor` call (which broke inter-graph tensor passing) into `compute_logits` where it becomes part of the compiled StableHLO graph. Remove the TP guard, external sharding constraint call, and `ParallelLMHead` isinstance check.
- **`test_sampling_params.py`**: Add `test_output_coherence` — push-tier test that verifies greedy output contains expected keywords for 6 simple prompts, catching gibberish regressions that the existing degenerate-token guard would miss. Issue #3663 has details.

### Checklist
- [x] All existing tests passing, nightly vLLM: https://github.com/tenstorrent/tt-xla/actions/runs/22969970036
- [x] New `test_output_coherence` passing before/after with no gibberish

### CODE REVIEW

**Comment by kmabeeTT** [comment]
### Design notes

**Why legacy compile for `sample_from_logits`?**

The experimental compile path (`torch.export` + `extract_compiled_graph`) fails on second invocation under SPMD for functions with no model parameters (#3672). `sample_from_logits` in greedy mode is just `argmax` — a pure function with no weights. Legacy compile (`tt_legacy_compile=True`) re-traces through torch_xla on each call, which works correctly. The overhead is host-side tracing on a very small graph. `gather_logprobs` doesn't need this workaround because it accesses `self.sampler` sub-modules, which changes the graph structure enough to avoid the bug.

**Why move sharding constraint inside `compute_logits`?**

The external `sharding_constraint_tensor` call between `compute_logits` and `sample_from_logits` created XLA lazy tensors that caused the second compiled function to produce incorrect results. Moving it inside `compute_logits` makes it part of the same compiled StableHLO graph. This is required for ParallelLMHead models because: (1) `LogitsProcessor._get_logits()` uses `quant_method.apply()` which bypasses `nn.Module.__call__` and therefore any forward hooks, and (2) `tensor_model_parallel_all_gather` is a no-op under TT SPMD (world_size=1). For VocabParallelEmbedding models (tied embeddings) the constraint is redundant but harmless — the compiler folds it out.

**Coherence test rationale**

During investigation, compiling `sample_from_logits` without proper logit replication produced deterministic but gibberish output (`"competition competition competitionsweisesweise"`). This passed `test_greedy_determinism` (non-degenerate, deterministic) but was clearly wrong. `test_output_coherence` catches this class of failure using 6 prompts that all test models reliably complete under greedy sampling.

File: integrations/vllm_plugin/vllm_tt/model_runner.py [hunk 1]
```diff
@@ -2008,11 +1986,20 @@ def select_hidden_states(self, hidden_states, indices_do_sample):
 
     @torch.compile(backend="tt", fullgraph=True, dynamic=False)
     def compute_logits(self, sample_hidden_states: torch.Tensor) -> torch.Tensor:
-        return self.model.compute_logits(sample_hidden_states)
+        logits = self.model.compute_logits(sample_hidden_states)
+        # Replicate logits for SPMD. Hooks can't reach ParallelLMHead
+        # (quant_method.apply bypasses __call__) and all_gather is a
+        # no-op (world_size=1). Must be inside the compiled graph —
+        # external sharding_constraint between compiled functions breaks.
+        if self.enable_tensor_parallel:
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: Regarding models testing - yeah n300 now uses TinyLlama for `ParallelLMHead` coverage since https://github.com/tenstorrent/tt-xla/pull/3653 yesterday, and the n300-llmbox specifically uses Qwen3-0.6B which has tied embeddings (`VocabParallelEmbedding`, not `ParallelLMHead`) to get coverage for both, and both are working here with these changes.

[the remark this answers is not in the document]
Comment (mmanzoorTT) [reply integrations/vllm_plugin/vllm_tt/model_runner.py]: Thanks for the explanation.
